### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [2]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data_df = pd.read_csv(file_to_load)

purchase_data_df.head(20)

Purchase ID             SN  Age                 Gender  Item ID  \
0             0        Lisim78   20                   Male      108   
1             1    Lisovynya38   40                   Male      143   
2             2     Ithergue48   24                   Male       92   
3             3  Chamassasya86   24                   Male      100   
4             4      Iskosia90   23                   Male      131   
5             5        Yalae81   22                   Male       81   
6             6      Itheria73   36                   Male      169   
7             7    Iskjaskst81   20                   Male      162   
8             8      Undjask33   22                   Male       21   
9             9    Chanosian48   35  Other / Non-Disclosed      136   
10           10      Inguron55   23                   Male       95   
11           11   Haisrisuir60   23                   Male      162   
12           12   Saelaephos52   21                   Male      116   
13           13    Assjaskan73   22                   Male        4   
14           14    Saesrideu94   35                   Male      165   
15           15      Lisassa64   21                 Female       98   
16           16      Lisirra25   20                   Male       40   
17           17      Zontibe81   21                   Male      161   
18           18      Reunasu60   22                 Female       82   
19           19      Chamalo71   30                   Male       89   

                                    Item Name  Price  
0   Extraction, Quickblade Of Trembling Hands   3.53  
1                           Frenzied Scimitar   1.56  
2                                Final Critic   4.88  
3                                 Blindscythe   3.27  
4                                        Fury   1.44  
5                                   Dreamkiss   3.61  
6      Interrogator, Blood Blade of the Queen   2.18  
7                               Abyssal Shard   2.67  
8                                   Souleater   1.10  
9                Ghastly Adamantite Protector   3.58  
10                      Singed Onyx Warscythe   4.74  
11                              Abyssal Shard   2.67  
12                    Renewed Skeletal Katana   4.18  
13                         Bloodlord's Fetish   1.70  
14                Bone Crushing Silver Skewer   4.86  
15                Deadline, Voice Of Subtlety   2.89  
16                              Second Chance   2.52  
17                                     Devine   1.76  
18                                    Nirvana   4.90  
19          Blazefury, Protector of Delusions   4.64

## Player Count

* Display the total number of players


In [3]:
players = {'Total Players':[len(purchase_data_df['SN'].value_counts())]}

total_players = pd.DataFrame(players)
total_players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
unique_items = purchase_data_df['Item ID'].nunique()
Average_price = purchase_data_df["Price"].mean()
total_purchases = purchase_data_df['Item Name'].count()
Total_revenue = purchase_data_df["Price"].sum()
dollars = "${0:.2f}".format
percentage = "{0:.2f}%".format


Display_Summary_data = [{"Number of Unique Items":unique_items, "Average Price":Average_price,
                         "Total Purchases":total_purchases,"Total Revenues":Total_revenue}]


Display_Summary_table = pd.DataFrame(Display_Summary_data)
Display_Summary_table["Average Price"] = Display_Summary_table["Average Price"].map(dollars)
Display_Summary_table["Total Revenues"] = Display_Summary_table["Total Revenues"].map(dollars)

Display_Summary_table

Number of Unique Items Average Price  Total Purchases Total Revenues
0                     183         $3.05              780       $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [6]:
grouped_gender_df = purchase_data_df.groupby(['Gender'])
total_unique_gender = grouped_gender_df['SN'].nunique()

gender_percentage = purchase_data_df

percentage_players = (total_unique_gender/[len(purchase_data_df['SN'].value_counts())])*100


summary_df = pd.DataFrame({"Total Count" : total_unique_gender, "Percentage of Players" : percentage_players })

summary_df["Percentage of Players"] = summary_df["Percentage of Players"].map(percentage)

summary_df

Total Count Percentage of Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
#grouped_gender_df = purchase_data_df.groupby(['Gender'])
grouped_gender_df.count()

total_purchases_gender = grouped_gender_df['Gender'].size()
average_purchases_gender = grouped_gender_df['Price'].mean().round(2)
total_revenue_gender = grouped_gender_df['Price'].sum()

Total_revenue_per_player = total_revenue_gender/total_unique_gender


summary_df_gender = pd.DataFrame({"Purchase Count" : total_purchases_gender, 
                                  "Average Purchase Price" : average_purchases_gender, 
                                  "Total Purchase Value" : total_revenue_gender,
                                  "Avg Total Purchase per Person" : Total_revenue_per_player})

summary_df_gender["Average Purchase Price"] = summary_df_gender["Average Purchase Price"].map(dollars)
summary_df_gender["Total Purchase Value"] = summary_df_gender["Total Purchase Value"].map(dollars)
summary_df_gender["Avg Total Purchase per Person"] = summary_df_gender["Avg Total Purchase per Person"].map(dollars)

summary_df_gender

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [8]:
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]

Group_names = ["<10","10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

purchase_data_df["Age groups"] = pd.cut(purchase_data_df["Age"], bins, labels=Group_names)
purchase_data_df_age = purchase_data_df.groupby("Age groups")

total_purchases_age = purchase_data_df_age['SN'].nunique()

total_purchases_age_percent = (purchase_data_df_age['SN'].nunique() / [len(purchase_data_df['SN'].value_counts())])*100

summary_df_age = pd.DataFrame({"Total Count" : total_purchases_age, 
                               "Percentage of Players" : total_purchases_age_percent})

summary_df_age["Percentage of Players"] = summary_df_age["Percentage of Players"].map(percentage)

summary_df_age

Total Count Percentage of Players
Age groups                                   
<10                  17                 2.95%
10-14                22                 3.82%
15-19               107                18.58%
20-24               258                44.79%
25-29                77                13.37%
30-34                52                 9.03%
35-39                31                 5.38%
40+                  12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [9]:
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]

Group_names = ["<10","10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

purchase_data_df["Age groups"] = pd.cut(purchase_data_df["Age"], bins, labels=Group_names)
purchase_data_df_age = purchase_data_df.groupby("Age groups")

total_purchases_age_count = purchase_data_df_age['Age'].size()
#total_purchases_age_count

average_purchases_age_count = purchase_data_df_age['Price'].mean().round(2)
#average_purchases_age_count

total_revenue_age_count = purchase_data_df_age['Price'].sum()
#total_revenue_age_count

Total_revenue_per_player_age_count = (total_revenue_age_count/[len(purchase_data_df_age['SN'].value_counts())])*100
#Total_revenue_per_player_age_count

summary_df_age_count = pd.DataFrame({"Purchase Count" : total_purchases_age_count, 
                               "Average Purchase Price" : average_purchases_age_count, 
                               "Total Purchase Value" : total_revenue_age_count, 
                               "Avg Total Purchase per Person" : Total_revenue_per_player_age_count})

summary_df_age_count["Average Purchase Price"] = summary_df_age_count["Average Purchase Price"].map(dollars)
summary_df_age_count["Total Purchase Value"] = summary_df_age_count["Total Purchase Value"].map(dollars)
summary_df_age_count["Avg Total Purchase per Person"] = summary_df_age_count["Avg Total Purchase per Person"].map(dollars)

summary_df_age_count

Purchase Count Average Purchase Price Total Purchase Value  \
Age groups                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05             $1114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age groups                                
<10                               $13.39  
10-14                             $14.37  
15-19                             $71.68  
20-24                            $193.41  
25-29                             $50.87  
30-34                             $37.15  
35-39                             $25.64  
40+                                $6.64

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [10]:

total_spender = purchase_data_df.groupby('SN')

total_spender_count = total_spender['Purchase ID'].count()

total_spender_average = total_spender['Price'].mean()

total_spender_revenue = total_spender['Price'].sum()


summary_df_spender = pd.DataFrame({"Purchase Count" : total_spender_count, 
                               "Average Purchase Price" : total_spender_average, 
                               "Total Purchase Value" : total_spender_revenue})

summary_df_table = summary_df_spender.sort_values('Total Purchase Value', ascending=False)

summary_df_table["Average Purchase Price"] = summary_df_table["Average Purchase Price"].map(dollars)
summary_df_table["Total Purchase Value"] = summary_df_table["Total Purchase Value"].map(dollars)

summary_df_table.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [13]:
most_popular = purchase_data_df.groupby(['Item ID', 'Item Name'])
most_popular.count()
most_popular_count = most_popular['Purchase ID'].count()

most_popular_average = most_popular['Price'].mean()

most_popular_revenue = most_popular['Price'].sum()


most_popular_spender = pd.DataFrame({"Purchase Count" : most_popular_count, 
                                     "Average Purchase Price" : most_popular_average, 
                                     "Total Purchase Value" : most_popular_revenue})

most_popular_table = most_popular_spender.sort_values('Purchase Count', ascending=False)

most_popular_table["Average Purchase Price"] = most_popular_table["Average Purchase Price"].map(dollars)
most_popular_table["Total Purchase Value"] = most_popular_table["Total Purchase Value"].map(dollars)

most_popular_table.head(20)

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   
103     Singed Scalpel                                             8   
75      Brutality Ivory Warmace                                    8   
72      Winter's Bite                                              8   
60      Wolf                                                       8   
59      Lightning, Etcher of the King                              8   
37      Shadow Strike, Glory of Ending Hope                        8   
34      Retribution Axe                                            8   
92      Final Critic                                               8   
53      Vengeance Cleaver                                          7   
110     Suspension                                                 7   
7       Thorn, Satchel of Dark Souls                               7   
71      Demise                                                     7   
117     Heartstriker, Legacy of the Light                          7   
159     Oathbreaker, Spellblade of Trials                          7   
85      Malificent Bag                                             7   

                                                     Average Purchase Price  \
Item ID Item Name                                                             
178     Oathbreaker, Last Hope of the Breaking Storm                  $4.23   
145     Fiery Glass Crusader                                          $4.58   
108     Extraction, Quickblade Of Trembling Hands                     $3.53   
82      Nirvana                                                       $4.90   
19      Pursuit, Cudgel of Necromancy                                 $1.02   
103     Singed Scalpel                                                $4.35   
75      Brutality Ivory Warmace                                       $2.42   
72      Winter's Bite                                                 $3.77   
60      Wolf                                                          $3.54   
59      Lightning, Etcher of the King                                 $4.23   
37      Shadow Strike, Glory of Ending Hope                           $3.16   
34      Retribution Axe                                               $2.22   
92      Final Critic                                                  $4.88   
53      Vengeance Cleaver                                             $2.05   
110     Suspension                                                    $1.44   
7       Thorn, Satchel of Dark Souls                                  $1.33   
71      Demise                                                        $1.61   
117     Heartstriker, Legacy of the Light                             $1.79   
159     Oathbreaker, Spellblade of Trials                             $3.08   
85      Malificent Bag                                                $1.75   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16  
103     Singed Scalpel                                             $34.80  
75      Brutality Ivory Warmace                                    $19.36  
72      Winter's Bite                         

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [14]:
most_popular = purchase_data_df.groupby(['Item ID', 'Item Name'])
most_popular.count()
most_popular_count = most_popular['Purchase ID'].count()

most_popular_average = most_popular['Price'].mean()

most_popular_revenue = most_popular['Price'].sum()


most_popular_spender = pd.DataFrame({"Purchase Count" : most_popular_count, 
                                     "Average Purchase Price" : most_popular_average, 
                                     "Total Purchase Value" : most_popular_revenue})

most_popular_table = most_popular_spender.sort_values('Total Purchase Value', ascending=False)

most_popular_table["Average Purchase Price"] = most_popular_table["Average Purchase Price"].map(dollars)
most_popular_table["Total Purchase Value"] = most_popular_table["Total Purchase Value"].map(dollars)

most_popular_table.head(20)

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   
59      Lightning, Etcher of the King                              8   
108     Extraction, Quickblade Of Trembling Hands                  9   
78      Glimmer, Ender of the Moon                                 7   
72      Winter's Bite                                              8   
60      Wolf                                                       8   
179     Wolf, Promise of the Moonwalker                            6   
160     Azurewrath                                                 6   
25      Hero Cane                                                  6   
37      Shadow Strike, Glory of Ending Hope                        8   
139     Mercy, Katana of Dismay                                    5   
128     Blazeguard, Reach of Eternity                              5   
138     Peacekeeper, Wit of Dark Magic                             5   
141     Persuasion                                                 7   
87      Deluge, Edge of the West                                   5   
159     Oathbreaker, Spellblade of Trials                          7   

                                                     Average Purchase Price  \
Item ID Item Name                                                             
178     Oathbreaker, Last Hope of the Breaking Storm                  $4.23   
82      Nirvana                                                       $4.90   
145     Fiery Glass Crusader                                          $4.58   
92      Final Critic                                                  $4.88   
103     Singed Scalpel                                                $4.35   
59      Lightning, Etcher of the King                                 $4.23   
108     Extraction, Quickblade Of Trembling Hands                     $3.53   
78      Glimmer, Ender of the Moon                                    $4.40   
72      Winter's Bite                                                 $3.77   
60      Wolf                                                          $3.54   
179     Wolf, Promise of the Moonwalker                               $4.48   
160     Azurewrath                                                    $4.40   
25      Hero Cane                                                     $4.35   
37      Shadow Strike, Glory of Ending Hope                           $3.16   
139     Mercy, Katana of Dismay                                       $4.94   
128     Blazeguard, Reach of Eternity                                 $4.91   
138     Peacekeeper, Wit of Dark Magic                                $4.74   
141     Persuasion                                                    $3.19   
87      Deluge, Edge of the West                                      $4.43   
159     Oathbreaker, Spellblade of Trials                             $3.08   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80  
59      Lightning, Etcher of the King                              $33.84  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
78      Glimmer, Ender of the Moon            

Observation from Data Sets

1) That Males represent 84% of the total 576 players.  Also 2% of the total survey group was not willing/able to state their gender (not sure why people would skip this question for this survey???)

2) That the age group of 20-24 had the largest percentage of players at 45% and this age group also spent the most money overall ($1114.06).  But the 35-39 age group spent the most money per game at a $3.60 average, possibly because this age group may make more money overall and willing to spend it on games.

3)The most popular game happened to be the most profitable but there are games that were very profitable (ex. Final Critic #4 on list) but not that popular overall (Final Critic #13 on popularity list).  This could mean that the game was interesting enough that people were willing to purchase it at a higher price so that compensated for the lack of popularity to keep profits high.